File: internalUsages/SharedDomain/ValueObjects/TemperatureCelsius.cs
using Orleans;

namespace SharedDomain.ValueObjects;

[GenerateSerializer]
public record TemperatureCelsius
{
    public int Value { get; init; }

    public TemperatureCelsius(int value)
    {
        if (value < -273)
        {
            throw new ArgumentException("Temperature cannot be below absolute zero.");
        }
        Value = value;
    }

    public static implicit operator TemperatureCelsius(int value) => new(value);
    public static implicit operator int(TemperatureCelsius temperature) => temperature.Value;

    public int ToFahrenheit() => 32 + (int)(Value / 0.5556);
}

File: ts/samples/dapr-sample/packages/domain/src/valueObjects/temperature-celsius.ts
import { z } from 'zod';

export const TemperatureCelsiusSchema = z.object({
  value: z.number().refine(
    (val) => val >= -273,
    { message: "Temperature cannot be below absolute zero." }
  )
});

export type TemperatureCelsius = z.infer<typeof TemperatureCelsiusSchema>;

export const createTemperatureCelsius = (value: number): TemperatureCelsius => {
  return TemperatureCelsiusSchema.parse({ value });
};

export const toFahrenheit = (temp: TemperatureCelsius): number => {
  return 32 + Math.floor(temp.value / 0.5556);
};